server/sonar-server/src/main/java/org/sonar/server/es/BulkIndexer.java

```diff
@@ -192,73 +154,139 @@ public void addDeletion(IndexType indexType, String id, String routing) {
-      if (semaphore.tryAcquire(concurrentRequests, 10, TimeUnit.MINUTES)) {
-        semaphore.release(concurrentRequests);
+
+    @Override
+    public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+      counter.addAndGet(response.getItems().length);
+
+      for (BulkItemResponse item : response.getItems()) {
+        if (item.isFailed()) {
+          LOGGER.error("index [{}], type [{}], id [{}], message [{}]", item.getIndex(), item.getType(), item.getId(), item.getFailureMessage());
+        }
       }
-    } catch (InterruptedException e) {
-      throw new IllegalStateException("Elasticsearch bulk requests still being executed after 10 minutes", e);
     }
-    progress.stop();
-    client.prepareRefresh(indexName).get();
-    if (size == Size.LARGE) {
-      // optimize lucene segments and revert index settings
-      // Optimization must be done before re-applying replicas:
-      // http://www.elasticsearch.org/blog/performance-considerations-elasticsearch-indexing/
-      client.prepareForceMerge(indexName).get();
 
-      updateSettings(largeInitialSettings);
+    @Override
+    public void afterBulk(long executionId, BulkRequest req, Throwable e) {
+      LOGGER.error("Fail to execute bulk index request: " + req, e);
+    }
+  }
+
+  public enum Size {
+    /** Use this size for a limited number of documents. */
+    REGULAR,
+
+    /**
+     * Large indexing is an heavy operation that populates an index generally from scratch. Replicas and
+     * automatic refresh are disabled during bulk indexing and lucene segments are optimized at the end.
+     * Use this size for initial indexing and if you expect unusual huge numbers of documents.
+     */
+    LARGE {
+      @Override
+      SizeHandler createHandler(Runtime2 runtime2) {
+        return new LargeSizeHandler(runtime2);
+      }
+    };
+
+    SizeHandler createHandler(Runtime2 runtime2) {
+      return new SizeHandler();
     }
-    bulkRequest = null;
   }
 
-  private void updateSettings(Map<String, Object> settings) {
-    UpdateSettingsRequestBuilder req = client.nativeClient().admin().indices().prepareUpdateSettings(indexName);
-    req.setSettings(settings);
-    req.get();
+  @VisibleForTesting
+  static class Runtime2 {
+    private static final Runtime2 INSTANCE = new Runtime2();
+
+    int getCores() {
+      return Runtime.getRuntime().availableProcessors();
+    }
   }
 
-  private void executeBulk() {
-    final BulkRequestBuilder req = this.bulkRequest;
-    this.bulkRequest = client.prepareBulk().setRefresh(false);
-    semaphore.acquireUninterruptibly();
-    req.execute(new BulkResponseActionListener(req));
+  static class SizeHandler {
+    /**
+     * @see BulkProcessor.Builder#setConcurrentRequests(int)
+     */
+    int getConcurrentRequests() {
+      // in the same thread by default
+      return 0;
+    }
+
+    void beforeStart(BulkIndexer bulkIndexer) {
+    }
+
+    void afterStop(BulkIndexer bulkIndexer) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1186)